tests/IntervalTest.php

```diff
@@ -95,6 +106,135 @@ public function testGetDuration(): void
+    public function testIntersectsWith(int $start1, int $end1, int $start2, int $end2, bool $expected): void
+    {
+        $interval1 = new Interval(Instant::of($start1), Instant::of($end1));
+        $interval2 = new Interval(Instant::of($start2), Instant::of($end2));
+        $this->assertEquals($expected, $interval1->intersectsWith($interval2));
+    }
+
+    public function providerIntersectsWith(): array
+    {
+        return [
+            'second is after first' => [
+                100000, 200000,
+                400000, 500000,
+                false,
+            ],
+            'second is before first' => [
+                400000, 500000,
+                100000, 200000,
+                false,
+            ],
+            'end of the first is start of the second' => [
+                100000, 200000,
+                200000, 300000,
+                false,
+            ],
+            'start of the first is end of the second' => [
+                200000, 300000,
+                100000, 200000,
+                false,
+            ],
+            'intersection' => [
+                100000, 200000,
+                150000, 250000,
+                true,
+            ],
+        ];
+    }
+
+    /** @dataProvider providerGetIntersectionWith */
+    public function testGetIntersectionWith(
+        int $start1,
+        int $end1,
+        int $start2,
+        int $end2,
+        int $expectedStart,
+        int $expectedEnd
+    ): void {
+        $interval1 = new Interval(Instant::of($start1), Instant::of($end1));
+        $interval2 = new Interval(Instant::of($start2), Instant::of($end2));
+        $expected = new Interval(Instant::of($expectedStart), Instant::of($expectedEnd));
+
+        $this->assertEquals($expected, $interval1->getIntersectionWith($interval2));
+    }
+
+    public function providerGetIntersectionWith(): array
+    {
+        return [
+            'first before second' => [
+                100000, 200000,
+                150000, 250000,
+                150000, 200000,
+            ],
+            'first after second' => [
+                150000, 250000,
+                100000, 200000,
+                150000, 200000,
+            ],
+            'first inside second' => [
+                200000, 300000,
+                100000, 400000,
+                200000, 300000,
+            ],
+            'second inside first' => [
+                100000, 400000,
+                200000, 300000,
+                200000, 300000,
+            ],
+            'first = second' => [
+                5000, 6000,
+                5000, 6000,
+                5000, 6000,
+            ],
+        ];
+    }
+
+    public function testGetIntersectionWithInvalidParams(): void
+    {
+        $interval1 = new Interval(Instant::of(100000), Instant::of(200000));
+        $interval2 = new Interval(Instant::of(300000), Instant::of(400000));
+        $this->expectException(DateTimeException::class);
```

Comment:
Can you add an expectation for the exception message?

Reply:
added